### PR TITLE
Launchpad: Improve task status implementation

### LINKED ISSUE #68466: LaunchPad: Follow up on task statuses
### Details

There has been an ongoing discussion about refactoring the [current implementation of the LaunchPad task statuses](https://github.com/Automattic/wp-calypso/blob/add/tracks_events_to_launchpad/client/landing/stepper/declarative-flow/internals/steps-repository/launchpad/checklist-item.tsx#L19-L23). The last time we worked on these was in #68052.

Currently, we have three task statuses:
`completed-and-active`, `completed-and-inactive` and `not-completed` where each of them is a combination of the `isCompleted` and `keepActive` props. This is arguably unnecessarily complicated and rather hard to read. Ideally, we would want to simplify the implementation and make it easier to understand. Possibly decreasing the number of css classes that we use.

### PR DESCRIPTION
**Pull request by sharpirate**
#### Proposed Changes
This PR improves simplifies the implementation for the task statuses and improves readability

#### Testing Instructions
All styling should remain unchanged from the current implementation on `wordpress.com`: chevron icons, hover styles, disabled tasks etc.

##### Newsletter Flow
1. Visit `http://calypso.localhost:3000/setup/launchpad?flow=newsletter&complete-setup=true&siteSlug=YOUR_SITE_SLUG`
2. **Personalize Newsletter** should be completed and disabled (can't be revisited)
3. **Choose a Plan** should be completed and enabled (can be revisited)
4. **Add Subscribers** should be completed and enabled (can be revisited)
5. **Write your first post** should be pending and enabled

##### LinkInBio Flow
1. Visit `http://calypso.localhost:3000/setup/launchpad?flow=link-in-bio&complete-setup=true&siteSlug=YOUR_SITE_SLUG`
2. **Select a design** should be completed and disabled (can't be revisited)
6. **Personalize Link In Bio** should be completed and enabled (can be revisited)
7. **Choose a Plan** should be completed and enabled (can be revisited)
8. **Add links** should be pending and enabled
9. **Launch Link in Bio** should be pending and disabled (due to incomplete dependencies)

#### Pre-merge Checklist

- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?

Fixes #68466

### CODE REVIEW

**Comment by sharpirate** [comment]
The gist of this PR is that we can have four unique statuses for a given task:

1. Completed and enabled (can be revisited)
2. Completed and disabled (can't be revisited)
3. Pending and enabled
4. Pending and disabled (due to incomplete dependent tasks)

We are deprecating the previous implementation with the `keepActive` prop which was another level of abstraction above whether a task was enabled or disabled (we still had separate logic for figuring that out with the `isTaskDisabled` function).

The new approach is that we have two properties on every task - `completed` and `disabled`. Any combination of those two can cover all of the 4 cases described above:

1. `completed: true`, `disabled: false`
2. `completed: true`, `disabled: true`
3. `completed: false`, `disabled: false`
4. `completed: false`, `disabled: true`

Since these are now built into each task we no longer need the `isTaskDisabled` function which was causing additional complexity. 

Another improvement is that we no longer need the `hasIncompleteDependencies` since we can control the `disabled` prop on each task directly and update it accordingly to the status of the dependent tasks.

We are also removing the `displayBadge` prop which was just a boolean which proved to be unnecessary because we can just check for the `badgeText` itself. As discussed there won't be a case where the `badgeText` will be null and the `displayBadge` will be true or vice versa

File: client/landing/stepper/declarative-flow/internals/steps-repository/launchpad/checklist-item.tsx
Comment (sharpirate) [inline client/landing/stepper/declarative-flow/internals/steps-repository/launchpad/checklist-item.tsx:18]: The `pending` and `enabled` css classes are just "syntactic sugar" for easier styling. Alternatively we had to use `:not()` selectors which increase the complexity and lower the readability

File: client/landing/stepper/declarative-flow/internals/steps-repository/launchpad/checklist-item.tsx
Comment (sharpirate) [inline client/landing/stepper/declarative-flow/internals/steps-repository/launchpad/checklist-item.tsx]: now the built-in `disabled` task prop integrates nicely with the actual disabled attribute of the HTML element (single source of truth) where previously we had the `keepActive` prop which was another abstraction

File: client/landing/stepper/declarative-flow/internals/steps-repository/launchpad/task-helper.ts
Comment (sharpirate) [inline client/landing/stepper/declarative-flow/internals/steps-repository/launchpad/task-helper.ts:173]: the `isTaskDisabled` is no longer necessary since we have access to the built-in `disabled` prop directly

File: client/landing/stepper/declarative-flow/internals/steps-repository/launchpad/task-helper.ts
Comment (sharpirate) [inline client/landing/stepper/declarative-flow/internals/steps-repository/launchpad/task-helper.ts:185]: the `hasIncompleteDependencies` is no longer necessary since we can modify the `disabled` prop directly, by checking if the task's dependencies are resolved (completed) and setting it to `false`, otherwise setting it to `true`

File: client/landing/stepper/declarative-flow/internals/steps-repository/launchpad/tasks.tsx
Comment (sharpirate) [inline client/landing/stepper/declarative-flow/internals/steps-repository/launchpad/tasks.tsx:6]: using the built-in `completed` and `disabled` props we can achieve every possible combination of a task's status

File: client/landing/stepper/declarative-flow/internals/steps-repository/launchpad/task-helper.ts
Comment (sharpirate) [inline client/landing/stepper/declarative-flow/internals/steps-repository/launchpad/task-helper.ts:105]: simplify the dependencies implementation by modifying the `disabled` prop directly

**Comment by sharpirate** [comment]
I would like to get feedback on the unit tests, I removed the ones that were tied to the deprecated helper functions `getIsTaskDisabled` and `hasIncompleteDependencies` and refactored the other tests to match the new properties. I am curious if there's an opportunity to add additional tests to complement the new implementation

**Comment by edanzer** [comment]
This tests correctly minus small issue below at bottom. Code changes seems straightforward. The logic was confusing before and this is much cleaner. Based on what I can tell, this still seems to cover all cases and situations correctly. 

**Newsletter Flow**
- [x] Personalize Newsletter should be completed and disabled (can't be revisited)
- [x] Choose a Plan should be completed and enabled (can be revisited)
- [x] Add Subscribers should be completed and enabled (can be revisited)
- [x] Write your first post should be pending and enabled

**LinkInBio Flow**
- [x] Select a design should be completed and disabled (can't be revisited)
- [x] Personalize Link In Bio should be completed and enabled (can be revisited)
- [x] Choose a Plan should be completed and enabled (can be revisited)
- [x] Add links should be pending and enabled
- [x] Launch Link in Bio should be pending and disabled (due to incomplete dependencies)
- [x] PLUS After adding links, Launch Link in Bio should be pending and enabled (dependencies now complete)

**Launchpad Tests**
- [x] All Launchpad tests still pass

Small issue:
- The Launch Link in Bio link has hover color (blue) even when disabled. I assume we don't want to show that?

**Comment by Addison-Stavlo** [comment]
> The Launch Link in Bio link has hover color (blue) even when disabled. I assume we don't want to show that?

definitely agree that doesn't seem right. Behaviorally it might be less of an issue when that last task is moved to a CTA button, but it seems like this might be highlighting a logical flaw here somewhere.

**Comment by jeyip** [comment]
## Testing
### Requirements
#### Newsletter
- [x]  Choose a Plan should be completed and enabled (can be revisited)
- [x]  Add Subscribers should be completed and enabled (can be revisited)
- [x]  Write your first post should be pending and enabled

#### Link in Bio
- [x] Select a design should be completed and disabled (can't be revisited)
- [x] Personalize Link In Bio should be completed and enabled (can be revisited)
- [x] Choose a Plan should be completed and enabled (can be revisited)
- [x] Add links should be pending and enabled
- [x] Launch Link in Bio should be pending and disabled (due to incomplete dependencies)
- [x] After adding links, Launch Link in Bio should be pending and enabled (dependencies now complete)

### Browsers
- [x] Chrome

I'm seeing the same issues that Erick described where The Launch Link in Bio link has hover color (blue) even when disabled.

**Comment by jeyip** [comment]
I think the hover color bug for the launch link in bio task is present on production as well 😬

**Comment by jeyip** [comment]
@Sharpirate To fix the error, we need to add styles that account for pending and disabled tasks, which, from what I can tell, we don't currently account for in `launchpad/style.scss`.

```
.launchpad__task.pending.disabled {
	.launchpad__checklist-item-text {
		color: var(--studio-gray-50); // unsure if this is the correct color or font-weight
	}
}
```

File: client/landing/stepper/declarative-flow/internals/steps-repository/launchpad/style.scss
Comment (sharpirate) [inline client/landing/stepper/declarative-flow/internals/steps-repository/launchpad/style.scss:335]: This fixes the hover color on the disabled pending tasks (launch Link In Bio)

Comment (edanzer) [reply client/landing/stepper/declarative-flow/internals/steps-repository/launchpad/style.scss:335]: Confirmed this is fixed.

File: client/landing/stepper/declarative-flow/internals/steps-repository/launchpad/task-helper.ts
Comment (edanzer) [inline client/landing/stepper/declarative-flow/internals/steps-repository/launchpad/task-helper.ts:37]: I think this probably got lost in rebasing. I updated keepActive to be true in [this PR](https://github.com/Automattic/wp-calypso/pull/68685). Personalize Newsletter is now an active link, which enabled users to edit their site title/tageline/icon on the post-setup page.

Since keepActive is no longer valid, we'd just want to change `disabled` to be false for this task in tasks.tsx. That will reenable the link here.

Comment (edanzer) [reply client/landing/stepper/declarative-flow/internals/steps-repository/launchpad/task-helper.ts:37]: This is resolved.

**Comment by edanzer** [comment]
Just retested this with recent changes. Everything looks good!

**Review by edanzer — APPROVED** [review]

File: client/landing/stepper/declarative-flow/internals/steps-repository/launchpad/test/checklist-item.tsx
Comment (jeyip) [inline client/landing/stepper/declarative-flow/internals/steps-repository/launchpad/test/checklist-item.tsx:56]: From what I can tell, these tests are still valid, because we still want to ensure behavior around a task that is dependent on other tasks...we've just refactored how we're approaching the problem. 

I think we should keep them and update the test implementation. Curious about your thoughts.

**Review by jeyip — COMMENTED** [review]
+1 from me when https://github.com/Automattic/wp-calypso/pull/69066/files#r999811191 is resolved

File: client/landing/stepper/declarative-flow/internals/steps-repository/launchpad/test/checklist-item.tsx
Comment (sharpirate) [reply client/landing/stepper/declarative-flow/internals/steps-repository/launchpad/test/checklist-item.tsx:56]: @jeyip Sorry for the late reply, I will look into this

**Comment by agrullon95** [comment]
### Testing:

**Newsletter Flow:** 

- [x] Personalize Newsletter should be completed and disabled (can be revisited - updated)
- [x] Choose a Plan should be completed and enabled (can be revisited)
- [x] Add Subscribers should be completed and enabled (can be revisited)
- [x] Write your first post should be pending and enabled

**Link-in-bio Flow:**

- [x] Select a design should be completed and disabled (can't be revisited)
- [x] Personalize Link In Bio should be completed and enabled (can be revisited)
- [x] Choose a Plan should be completed and enabled (can be revisited)
- [x] Add links should be pending and enabled
- [x] Launch Link in Bio should be pending and disabled (due to incomplete dependencies)

I think we should keep the removed test cases since they are still relevant. We want to test the same behavior for the checklist tasks: https://github.com/Automattic/wp-calypso/pull/69066/files#diff-c1574bb61f7c32711cb6f8e19242c2af6546d41f576e8894644783599c013135L56.

Will approve after merge conflicts are fixed and test case changes are added (if you think they're relevant or not).

---

Retested with the same instructions above and the changes look good.

**Review by jeyip — APPROVED** [review]

**Comment by edanzer** [comment]
There have been a range of changes since I last tested, so just running through once more now. Everything looks good. 

**Newsletter Flow**
- [x] Personalize Newsletter should be completed and enabled (can be revisited)
  - Note testing instructions are wrong on this point, as this has changed since the initial PR. It was enabled, and included here via rebase. 
- [x] Choose a Plan should be completed and enabled (can be revisited)
- [x] Add Subscribers should be completed and enabled (can be revisited)
- [x] Write your first post should be pending and enabled

**LinkInBio Flow**
- [x] Select a design should be completed and disabled (can't be revisited)
- [x] Personalize Link In Bio should be completed and enabled (can be revisited)
- [x] Choose a Plan should be completed and enabled (can be revisited)
- [x] Add links should be pending and enabled
- [x] Launch Link in Bio should be pending and disabled (due to incomplete dependencies)
- [x] After adding links, Launch Link in Bio should be pending and enabled (dependencies now complete)

**Launchpad Tests**
- [x] All Launchpad tests still pass

**Review by agrullon95 — APPROVED** [review]

**Pull request merged**
